File: app/src/main/java/com/zeyad/generic/genericrecyclerview/adapter/screens/user/list/UserListVM.java
package com.zeyad.generic.genericrecyclerview.adapter.screens.user.list;

import com.zeyad.generic.genericrecyclerview.adapter.screens.user.list.events.DeleteUsersEvent;
import com.zeyad.generic.genericrecyclerview.adapter.screens.user.list.events.GetPaginatedUsersEvent;
import com.zeyad.generic.genericrecyclerview.adapter.screens.user.list.events.SearchUsersEvent;
import com.zeyad.rxredux.core.redux.BaseEvent;
import com.zeyad.rxredux.core.redux.BaseViewModel;
import com.zeyad.rxredux.core.redux.SuccessStateAccumulator;
import com.zeyad.usecases.api.IDataService;
import com.zeyad.usecases.requests.GetRequest;

import java.util.List;

import io.reactivex.Flowable;
import io.reactivex.annotations.NonNull;
import io.reactivex.functions.Function;

import static com.zeyad.generic.genericrecyclerview.adapter.screens.utils.Constants.URLS.USERS;

/**
 * @author zeyad on 11/1/16.
 */
public class UserListVM extends BaseViewModel<UserListState> {

    private IDataService dataUseCase;

    @Override
    public void init(SuccessStateAccumulator<UserListState> successStateAccumulator,
            UserListState initialState, Object... otherDependencies) {
        dataUseCase = (IDataService) otherDependencies[0];
        setSuccessStateAccumulator(successStateAccumulator);
        setInitialState(initialState);
    }

    @Override
    public Function<BaseEvent, Flowable<?>> mapEventsToExecutables() {
        return new Function<BaseEvent, Flowable<?>>() {
            @Override
            public Flowable<?> apply(@NonNull BaseEvent event) throws Exception {
                Flowable executable = Flowable.empty();
                if (event instanceof GetPaginatedUsersEvent) {
                    executable = getUsers(((GetPaginatedUsersEvent) event).getLastId());
                } else if (event instanceof DeleteUsersEvent) {
                    //                    executable = deleteCollection(((DeleteUsersEvent) event).getSelectedItemsIds());
                } else if (event instanceof SearchUsersEvent) {
                    //                    executable = search(((SearchUsersEvent) event).getQuery());
                }
                return executable;
            }
        };
    }

    //    public Flowable<User> getUser() {
    //        return dataUseCase.getObjectOffLineFirst(new GetRequest.Builder(User.class, true)
    //                .url(String.format(USER, "Zeyad-37"))
    //                .id("Zeyad-37", User.LOGIN, String.class)
    //                .cache(User.LOGIN)
    //                .build());
    //    }
    //
    public Flowable<List<User>> getUsers(long lastId) {
        //        return lastId == 0 ?
        //                dataUseCase.getListOffLineFirst(new GetRequest.Builder(User.class, false)
        //                        .url(String.format(USERS, lastId))
        //                        .cache(User.LOGIN)
        //                        .build()) :
        return dataUseCase.getList(new GetRequest.Builder(User.class, false)
                .url(String.format(USERS, lastId))
                .build());
    }
    //
    //    public Flowable<List<User>> search(String query) {
    //        return dataUseCase.<User>queryDisk(realm -> realm.where(User.class).beginsWith(User.LOGIN, query))
    //                .zipWith(dataUseCase.<User>getObject(new GetRequest.Builder(User.class, false)
    //                                .url(String.format(USER, query))
    //                                .build())
    //                                .onErrorReturnItem(new User())
    //                                .filter(user -> user.getId() != 0)
    //                                .map(user -> user != null ?
    //                                        Collections.singletonList(user) : Collections.emptyList()),
    //                        (BiFunction<List<User>, List<User>, List<User>>) (users, singleton) -> {
    //                            users.addAll(singleton);
    //                            return new ArrayList<>(new HashSet<>(users));
    //                        });
    //    }
    //
    //    public Flowable<List<String>> deleteCollection(List<String> selectedItemsIds) {
    //        return dataUseCase.deleteCollectionByIds(new PostRequest.Builder(User.class, true)
    //                .payLoad(selectedItemsIds)
    //                .idColumnName(User.LOGIN, String.class)
    //                .cache()
    //                .queuable(false, false)
    //                .build())
    //                .map(o -> selectedItemsIds);
    //    }
}

File: app/src/main/java/com/zeyad/app/screens/list/UserListVM.kt
package com.zeyad.app.screens.list

import androidx.recyclerview.widget.DiffUtil
import com.zeyad.app.R
import com.zeyad.app.screens.User
import com.zeyad.app.screens.UserDiffCallBack
import com.zeyad.app.utils.Constants.URLS.USER
import com.zeyad.app.utils.Constants.URLS.USERS
import com.zeyad.gadapter.ItemInfo
import com.zeyad.rxredux.core.BaseEvent
import com.zeyad.rxredux.core.StringMessage
import com.zeyad.rxredux.core.viewmodel.BaseViewModel
import com.zeyad.usecases.api.IDataService
import com.zeyad.usecases.db.RealmQueryProvider
import com.zeyad.usecases.requests.GetRequest
import com.zeyad.usecases.requests.PostRequest
import io.reactivex.Flowable
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.functions.BiFunction
import io.realm.Realm
import io.realm.RealmQuery

class UserListVM(private val dataUseCase: IDataService) : BaseViewModel<UserListState, UserListEffect>() {
    override var disposable: CompositeDisposable = CompositeDisposable()

    override fun errorMessageFactory(throwable: Throwable, event: BaseEvent<*>) =
            StringMessage(throwable.localizedMessage)

    override fun mapEventsToActions(event: BaseEvent<*>): Flowable<*> {
        val userListEvent = event as UserListEvents<*>
        return when (userListEvent) {
            is GetPaginatedUsersEvent -> getUsers(userListEvent.getPayLoad())
            is DeleteUsersEvent -> deleteCollection(userListEvent.getPayLoad())
            is SearchUsersEvent -> search(userListEvent.getPayLoad())
        }
    }

    override fun reducer(newResult: Any, event: BaseEvent<*>, currentStateBundle: UserListState): UserListState {
        val currentItemInfo = currentStateBundle.list.toMutableList()
        return when (currentStateBundle) {
            is EmptyState -> when (newResult) {
                is List<*> -> {
                    val pair = Flowable.fromIterable(newResult as List<User>)
                            .map { ItemInfo(it, R.layout.user_item_layout, it.id) }
                            .toList()
                            .map { it as MutableList<ItemInfo<*>> }
                            .toFlowable()
                            .calculateDiff(currentItemInfo)
                    GetState(pair.first, pair.first[pair.first.size - 1].id, pair.second)
                }
                else -> throw IllegalStateException("Can not reduce GetState with this result: $newResult!")
            }
            is GetState -> when (newResult) {
                is List<*> -> {
                    val pair = Flowable.fromIterable(newResult as List<User>)
                            .map { ItemInfo(it, R.layout.user_item_layout, it.id) }
                            .toList()
                            .map {
                                val list = currentStateBundle.list.toMutableList()
                                list.addAll(it)
                                list.toSet().toMutableList()
                            }.toFlowable()
                            .calculateDiff(currentItemInfo)
                    GetState(pair.first, pair.first[pair.first.size - 1].id, pair.second)
                }
                else -> throw IllegalStateException("Can not reduce GetState with this result: $newResult!")
            }
        }
    }

    private fun Flowable<MutableList<ItemInfo<*>>>.calculateDiff(initialList: MutableList<ItemInfo<*>>)
            : Pair<MutableList<ItemInfo<*>>, DiffUtil.DiffResult> =
            scan(Pair(initialList, DiffUtil.calculateDiff(UserDiffCallBack(mutableListOf(), mutableListOf()))))
            { pair1, next ->
                Pair(next, DiffUtil.calculateDiff(UserDiffCallBack(pair1.first, next)))
            }.skip(1)
                    .blockingFirst()

    private fun getUsers(lastId: Long): Flowable<List<User>> {
//        return if (lastId == 0L)
//            dataUseCase.getListOffLineFirst(GetRequest.Builder(User::class.java, true)
//                    .url(String.format(USERS, lastId))
//                    .build())
//        else
        return dataUseCase.getList(GetRequest.Builder(User::class.java, true)
                .url(String.format(USERS, lastId)).build())
    }

    private fun search(query: String): Flowable<List<User>> {
        return dataUseCase
                .queryDisk(object : RealmQueryProvider<User> {
                    override fun create(realm: Realm): RealmQuery<User> =
                            realm.where(User::class.java).beginsWith(User.LOGIN, query)
                })
                .zipWith(dataUseCase.getObject<User>(GetRequest.Builder(User::class.java, false)
                        .url(String.format(USER, query)).build())
                        .onErrorReturnItem(User())
                        .filter { user -> user.id != 0L }
                        .map { mutableListOf(it) },
                        BiFunction<List<User>, MutableList<User>, List<User>>
                        { singleton, users ->
                            users.addAll(singleton)
                            users.asSequence().toSet().toList()
                        })
    }

    private fun deleteCollection(selectedItemsIds: List<String>): Flowable<List<String>> {
        return dataUseCase.deleteCollectionByIds<Any>(PostRequest.Builder(User::class.java, true)
                .payLoad(selectedItemsIds)
                .idColumnName(User.LOGIN, String::class.java).cache()
                .build())
                .map { selectedItemsIds }
    }
}
